File: modules/react-native-nitro-recycler-list/ios/RecyclerTabCoordinatorRegistry.swift
import Foundation

final class RecyclerTabCoordinatorState {
  private var offsets: [String: Double] = [:]
  private(set) var collapseOffset: Double = 0

  func update(tabKey: String, offset: Double, active: Bool, collapseRange: Double) {
    offsets[tabKey] = max(0, offset)
    if active { collapseOffset = min(max(0, collapseRange), max(0, offset)) }
  }

  func targetOffset(tabKey: String, collapseRange: Double) -> Double {
    let collapsed = min(max(0, collapseRange), collapseOffset)
    let saved = offsets[tabKey] ?? 0
    return collapsed < collapseRange ? collapsed : max(collapsed, saved)
  }
}

private final class WeakRecyclerList {
  weak var value: HybridRecyclerListView?
  init(_ value: HybridRecyclerListView) { self.value = value }
}

/// 用弱引用关联折叠 Tab 内的列表，并保存各页偏移。
final class RecyclerTabCoordinatorRegistry {
  private static var states: [String: RecyclerTabCoordinatorState] = [:]
  private static var lists: [String: [String: WeakRecyclerList]] = [:]
  private static let lock = NSLock()

  static func register(_ list: HybridRecyclerListView) {
    guard !list.tabCoordinatorId.isEmpty, !list.tabKey.isEmpty else { return }
    lock.lock()
    lists[list.tabCoordinatorId, default: [:]][list.tabKey] = WeakRecyclerList(list)
    if states[list.tabCoordinatorId] == nil { states[list.tabCoordinatorId] = RecyclerTabCoordinatorState() }
    lock.unlock()
  }

  static func unregister(_ list: HybridRecyclerListView) {
    unregister(list, coordinatorId: list.tabCoordinatorId, tabKey: list.tabKey)
  }

  static func unregister(_ list: HybridRecyclerListView, coordinatorId: String, tabKey: String) {
    lock.lock()
    if lists[coordinatorId]?[tabKey]?.value === list {
      lists[coordinatorId]?.removeValue(forKey: tabKey)
    }
    if lists[coordinatorId]?.isEmpty == true {
      lists.removeValue(forKey: coordinatorId)
      states.removeValue(forKey: coordinatorId)
    }
    lock.unlock()
  }

  static func update(_ list: HybridRecyclerListView, offset: Double) {
    guard !list.tabCoordinatorId.isEmpty, !list.tabKey.isEmpty else { return }
    lock.lock()
    let state = states[list.tabCoordinatorId] ?? RecyclerTabCoordinatorState()
    states[list.tabCoordinatorId] = state
    state.update(tabKey: list.tabKey, offset: offset, active: list.tabActive, collapseRange: list.tabCollapseRange)
    lock.unlock()
  }

  static func targetOffset(for list: HybridRecyclerListView) -> Double {
    lock.lock()
    let target = states[list.tabCoordinatorId]?.targetOffset(tabKey: list.tabKey, collapseRange: list.tabCollapseRange) ?? 0
    lock.unlock()
    return target
  }
}

File: modules/react-native-nitro-recycler-list/android/src/main/java/com/margelo/nitro/recyclerlist/RecyclerTabCoordinatorRegistry.kt
package com.margelo.nitro.recyclerlist

import java.lang.ref.WeakReference
import kotlin.math.max
import kotlin.math.min

internal class RecyclerTabCoordinatorState {
  private val offsets = HashMap<String, Double>()
  var collapseOffset: Double = 0.0
    private set

  fun update(tabKey: String, offset: Double, active: Boolean, collapseRange: Double) {
    offsets[tabKey] = max(0.0, offset)
    if (active) collapseOffset = min(max(0.0, collapseRange), max(0.0, offset))
  }

  fun targetOffset(tabKey: String, collapseRange: Double): Double {
    val collapsed = min(max(0.0, collapseRange), collapseOffset)
    val saved = offsets[tabKey] ?: 0.0
    return if (collapsed < collapseRange) collapsed else max(collapsed, saved)
  }
}

/** 用弱引用关联同一折叠 Tab 中的列表，并保存各页原生偏移。 */
internal object RecyclerTabCoordinatorRegistry {
  private val states = HashMap<String, RecyclerTabCoordinatorState>()
  private val lists = HashMap<String, MutableMap<String, WeakReference<HybridRecyclerListView>>>()

  @Synchronized
  fun register(list: HybridRecyclerListView) {
    if (list.tabCoordinatorId.isEmpty() || list.tabKey.isEmpty()) return
    lists.getOrPut(list.tabCoordinatorId) { HashMap() }[list.tabKey] = WeakReference(list)
    states.getOrPut(list.tabCoordinatorId) { RecyclerTabCoordinatorState() }
  }

  @Synchronized
  fun unregister(list: HybridRecyclerListView) {
    unregister(list.tabCoordinatorId, list.tabKey, list)
  }

  @Synchronized
  fun unregister(coordinatorId: String, tabKey: String, list: HybridRecyclerListView) {
    lists[coordinatorId]?.let { group ->
      if (group[tabKey]?.get() === list) group.remove(tabKey)
      if (group.isEmpty()) {
        lists.remove(coordinatorId)
        states.remove(coordinatorId)
      }
    }
  }

  @Synchronized
  fun update(list: HybridRecyclerListView, offset: Double) {
    if (list.tabCoordinatorId.isEmpty() || list.tabKey.isEmpty()) return
    states.getOrPut(list.tabCoordinatorId) { RecyclerTabCoordinatorState() }
      .update(list.tabKey, offset, list.tabActive, list.tabCollapseRange)
  }

  @Synchronized
  fun targetOffset(list: HybridRecyclerListView): Double =
    states[list.tabCoordinatorId]?.targetOffset(list.tabKey, list.tabCollapseRange) ?: 0.0
}
